### ISSUE TITLE
VBCSCompiler service remains in memory after `dotnet build` and consumes 100% of 1 CPU core and a lot of memory on Ubuntu

### ISSUE THREAD

**Issue by livarcocc** [dotnet/roslyn#27566]
_From @NikolayPianikov on June 7, 2018 12:40_

## Steps to reproduce

- create a temp directory with long full name (~>100 symbols)
- set the environment variable `TMPDIR` as full path to this temp directory
- build some dotnet project

⚠️ the memory runs out quickly and only the computer restart helps in some cases

[Sample repo](https://github.com/NikolayPianikov/dotnetcli_tpmdir_issue)

## Expected behavior

At least do not consume CPU and memory so hard

## Actual behavior

The process like

`/usr/share/dotnet/dotnet/usr/share/dotnet/sdk/2.1.300/Roslyn/bincore/VBCSCompiler.dll -pipename: xxx.F.DCgnN5efPzMYmbzHdZB6DrTe`

consumes 100% of 1 CPU core and a lot of memory

## Environment data

dotnet --version: 2.1.300
Ubuntu 18.04 LTS

_Copied from original issue: dotnet/cli#9434_

**Comment by livarcocc** [comment]
_From @zone117x on June 7, 2018 15:28_

Surprised to see this is a core problem with the dotnet cli. Definitely thought it was a bug with the Team City runner. 
Any idea why this isn't being reported much more frequently? Is the pipename/service archecture that the Teamcity buildrunner use a unique method? 

Since this will probably take a while to get fixed and stable, I'm very interested in any workarounds. I tried setting /nodereuse=false in my build steps but the process spawns and sticks anyway.

To give more bug report context: this issue has existed since the previous dotnet SDK stable version from a month ago (which I guess the version is now labeled `2.0 SDK (v2.1.200)`. And its happening on Ubuntu 16.04.

**Comment by livarcocc** [comment]
You can turn off the VBCSCompiler by setting UseSharedCompilation=false. You can do that as a parameter to the build command, you can set that as a MSBuild property in your projects or you can set that as an environment variable.

You can also use the `dotnet build-server shutdown` command to stop any long running servers spawned by the CLI once your build is done.

I will also move this issue to Roslyn to be investigated there.

**Comment by zone117x** [comment]
Running `dotnet build-server shutdown` has no effect on the spinning process. It remains running, even though the command seems to think it was successful:
```
Shutting down MSBuild server...
Shutting down VB/C# compiler server...
MSBuild server shut down successfully.
VB/C# compiler server shut down successfully.
```

**Comment by zone117x** [comment]
I added the following build agent environment variables which seems to work. I'm assuming only one is necessary but didn't test..
```
env.MSBUILDDISABLENODEREUSE=1
env.MSBUILDUSESHAREDCOMPILATION=false
env.UseSharedCompilation=false
```

**Comment by NikolayPianikov** [comment]
@zone117x I am going to add one of them to as default for dotnet cli plugin

**Comment by jcouv** [comment]
@livarcocc To investigate, I'd want to see what that process is doing. But I'm not sure how to do that on Linux/Ubuntu. Could we capture a debuggable memory dump? Do you have any pointers?

**Comment by jaredpar** [comment]
@agocke is this a duplicate of the pipe length issue that we fixed?

**Comment by agocke** [comment]
Yup, this looks like a duplicate of https://github.com/dotnet/roslyn/issues/24137
